git/Cheatsheat.md

```diff
@@ -0,0 +1,98 @@
+
+- Don't put **punctuation marks** at the end of commit message.
+
+    **GOOD:** `Remove unused code`
+
+    **WRONG:** ~~`Remove unused code.`~~
+
+- Use short commit messages - up to 50 characters.
+
+    **GOOD:** `Remove unused code`
+
+    **WRONG:** ~~`In this commit we've removed unused code along with classes`~~
+
+- Don't put more statements in one commit message. Use description instead. Description should be wrapped to 80 characters maximum.
+
+    **GOOD:**
+    ```
+    Remove unused code
+
+    Removed 'getAllUsers' function from 'UserDAO' class.
+    Performed cleanup on 'Users' controller. Removed unused
+    config file.
+    ```
+
+    **WRONG:** ~~`Remove 'getAllUsers' function. Cleanup 'Users' controller. Remove unused config.`~~
+
+- Leave second line blank when writing descriptions.
+
+    **GOOD:**
+    ```
+    Remove unused code
+
+    Removed 'getAllUsers' function from 'UserDAO' class.
+    Performed cleanup on 'Users' controller. Removed unused
+    config file.
+    ```
+
+    **WRONG:**
+    ```
+    Remove unused code
+    Removed 'getAllUsers' function from 'UserDAO' class.
+    Performed cleanup on 'Users' controller. Removed unused
+    config file.
+    ```
+
+### Useful Git commands and how to use them
+
+We assume that you've already known basic Git commands. So, we won't list you basic commands. We'll focus on specific commands that could be useful sometimes.
+
+1. `git commit --amend`
+    This command is used when we don't want to make two commits with same purpose. For example, if we're push some code and we notice that we forget to delete `print` from our code, then there's no sense to make another commit with `Remove unused 'print' command`. Then, we just remove it and use `--amend` flag instead.
+2. `git reset HEAD~`
+    This is used to **undo** your last commit locally. If you type `git commit -m "Implement something"` and you notice you've made mistake, then you can use this command to reset your latest work.
+
+### How to rebase properly?
+
+Rebasing technique is used when another branch is merged before your. Then, because we wan't to avoid `Git default merge` commits when you're **resolving conflicts**. Instead, we'll perform process to **put** your branch on updated `dev` instead.
+
+Assume that we have two branches `branch-1` and `branch-2`:
+
+```
+git checkout -b branch-1 dev
+git checkout -b branch-2 dev
+```
+
+And, we performed some work on `branch-2`, but `branch-1` is merged to `dev` in the meantime. Then we need to **rebase** our `branch-2`. To do that, type following commands:
+
+```
+git checkout dev
+git pull origin dev --rebase
+git checkout branch-2
+git rebase dev
+```
+
+Now, we can face three different scenarios:
+
+1. Desired one. There's no conflicts, and our branch is successfully rebased.
+2. We have conflict and there are conflicted files in our **index** with flag **both modified**. Then we modify those files, and type `git rebase --continue`.
```

Comment:
Maybe, it should be useful to mention adding files to index before `git rebase --continue` command.